### PR TITLE
Fix project.sync to allow configure preserve

### LINKED ISSUE #18177: `project.sync` is not allowing to specify `preserve` even though the `Sync` task does and the documentation explicitly mentions it
[`Project.sync`](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#sync-org.gradle.api.Action-) is documented to behave equal to the `Sync` task, explictely mentioning:

> Note that you can preserve output that already exists in the destination directory:
> ```
>  project.sync {
>      from 'source'
>      into 'dest'
>      preserve {
>          include 'extraDir/**'
>          include 'dir1/**'
>          exclude 'dir1/extra.txt'
>      }
>  }
> ```

This is not true. `Project.sync` only accepts a `CopySpec`, which does not allow to specify `preserve` patterns.

### Expected Behavior
`Project.sync` to allow specifying the same options as in the `Sync` task, e.g. by accepting a (new) `SyncSpec` instead of a `CopySpec`.

### Current Behavior
The `preserve` option of `Project.sync` is not available.

### Context
Renaming files in a directory, by preserving the files that already had been renamed.

### Steps to Reproduce 
Try calling the command that is documented in the JavaDoc of `Project.sync`:
```
project.sync {
    from 'source'
    into 'dest'
    preserve {
        include 'extraDir/**'
        include 'dir1/**'
        exclude 'dir1/extra.txt'
    }
}
```

### Your Environment
Every environment.

### PR DESCRIPTION
**Pull request by asodja**
This is a draft to fix #18177. Some more integration tests have to be added.

The problem was that project.sync accepted action of `CopySpec`and did allow to have preserve. This PR adds new interface `SyncSpec` and with that option to use preserve with `project.sync()` method.

There are some internal classes that that implement CopySpec:
- DefaultCopySpec
- CopySpecWrapper
- CopySpecInternal
- DelegatingCopySpecInternal

We decided for now that we should just change these to implement SyncSpec. But it's might be better from architecture point of view to create new classes, e.g.:
- DefaultSyncSpec
- SyncSpecWrapper
- SyncSpecInternal
- DelegatingSnycSpecInternal

What do you think?

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by big-guy** [comment]
@bot-gradle help

**Comment by big-guy** [comment]
@bot-gradle test AST

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    @Internal
+    PatternFilterable getPreserve();
+
+    /**
+     * Configures the filter that defines which files to preserve in the destination directory.
+     *
+     * @param action Action for configuring the preserve filter
+     * @return this
+     *
```
Comment (donat) [inline subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java:47]:
```suggestion
     * @return this
```

File: subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ */
+@Incubating
+@HasInternalProtocol
+public interface SyncSpec extends CopySpec {
+
+    /**
+     * Returns the filter that defines which files to preserve in the destination directory.
+     *
+     * @return the filter defining the files to preserve
+     *
```
Comment (donat) [inline subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java:37]:
```suggestion
     * @return the filter defining the files to preserve
```

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy [hunk 3]
```diff
@@ -328,14 +337,8 @@ class SyncTaskIntegrationTest extends AbstractIntegrationSpec {
...
-        file('dest').assertHasDescendants(
-            'dir1/file1.txt',
-            'dir2/subdir/file2.txt',
-            'dir2/file3.txt',
-            'emptyDir'
-        )
-        file('dest/emptyDir').exists()
-        !file('dest/extra1.txt').exists()
+        file('dest/dir1/extra1.txt').exists()
+        file('dest/extraDir/extra1.txt').exists()
```
Comment (donat) [inline subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy:341]: ❌ We accidentally removed some assertions.

```suggestion
        file('dest').assertHasDescendants(
            'dir1/file1.txt',
            'dir2/subdir/file2.txt',
            'dir2/file3.txt',
            'emptyDir'
            'dir1/extra1.txt',
            'extraDir/extra1.txt'
        )
```

Alternatively, we could restore the existing test and create a separate one with a `preserve {}` block.

(note: I've added the suggestion by hand; it needs testing)

**Review by donat — COMMENTED** [review]
I just found a few formatting errors and one thing to fix in the test coverage. Once fixed, I think we should be good for a review.

**Comment by donat** [comment]
@bot-gradle test rFM

File: subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java (hunk 1 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java]: done

File: subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java (hunk 2 above)
Comment (asodja) [reply subprojects/core-api/src/main/java/org/gradle/api/file/SyncSpec.java]: done

File: subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy (hunk 3 above)
Comment (asodja) [reply subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/SyncTaskIntegrationTest.groovy]: Readded old test and applied suggestions.

**Comment by asodja** [comment]
@bot-gradle test rFM

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle help

**Comment by big-guy** [comment]
@bot-gradle test AST

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@big-guy is this one good to be merged?

**Review by big-guy — APPROVED** [review]

**Comment by asodja** [comment]
@bot-gradle test and merge

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test and merge

**Pull request merged**